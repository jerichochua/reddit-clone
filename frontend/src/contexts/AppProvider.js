import React, { createContext, useContext, useReducer } from 'react';
import jwt_decode from 'jwt-decode';

const AppContext = createContext();
AppContext.displayName = 'AppContext';

const token = localStorage.getItem('token');
const userId = token ? jwt_decode(token).id : null;
const username = token ? jwt_decode(token).username : null;

const initialState = {
  userId,
  username,
  token,
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_USER':
      return {
        ...state,
        userId: action.payload.userId,
        username: action.payload.username,
      };
    case 'SET_TOKEN':
      return { ...state, token: action.payload };
    case 'LOGOUT':
      return { ...state, userId: null, username: null, token: null };
    default:
      return state;
  }
};

export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <AppContext.Provider value={{ state, dispatch }}>
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = () => {
  return useContext(AppContext);
};
