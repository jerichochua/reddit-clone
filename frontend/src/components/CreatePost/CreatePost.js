import React, { useEffect, useState } from 'react';
import { post } from '../../util/api';
import {
  Error,
  Form,
  FormButton,
  FormField,
  FormOption,
  FormTextArea,
} from '../Form';
import { useAppContext } from '../../contexts/AppProvider';
import {
  validateContent,
  validateTitle,
  validateUrl,
} from '../../util/validators';

const CreatePost = () => {
  const { state } = useAppContext();
  const [titleErrors, setTitleErrors] = useState([]);
  const [contentErrors, setContentErrors] = useState([]);
  const [urlErrors, setUrlErrors] = useState([]);
  const [currentPostType, setCurrentPostType] = useState('text');

  const titleErrorComponents = titleErrors.map((error, index) => (
    <Error key={index} message={error} />
  ));

  const contentErrorComponents = contentErrors.map((error, index) => (
    <Error key={index} message={error} />
  ));

  const urlErrorComponents = urlErrors.map((error, index) => (
    <Error key={index} message={error} />
  ));

  const onSubmitForm = async (e) => {
    e.preventDefault();

    setTitleErrors([]);
    setContentErrors([]);
    setUrlErrors([]);

    let body = {};
    const titleErrors = validateTitle(e.target.title.value);
    if (titleErrors.length > 0) {
      setTitleErrors(titleErrors);
      return;
    }

    if (currentPostType === 'link') {
      const urlErrors = validateUrl(e.target.url.value);
      if (urlErrors.length > 0) {
        setUrlErrors(urlErrors);
        return;
      }
      body = {
        title: e.target.title.value,
        type: 'link',
        url: e.target.url.value,
      };
    } else {
      const contentErrors = validateContent(e.target.content.value);
      if (contentErrors.length > 0) {
        setContentErrors(contentErrors);
        return;
      }
      body = {
        title: e.target.title.value,
        type: 'text',
        content: e.target.content.value,
      };
    }

    try {
      const response = await post('posts', body, state.token);
      if (response) {
        window.location.href = `/posts/${response.id}`;
      } else {
        console.error(response);
      }
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (!state.token) {
      window.location.href = '/login';
    }
  }, [state.token]);

  return (
    <Form wide onSubmit={onSubmitForm}>
      <FormOption
        name='posttype'
        label='type'
        options={[
          { value: 'text', label: 'text' },
          { value: 'link', label: 'link' },
        ]}
        setOption={setCurrentPostType}
      />
      <FormField
        name='title'
        label='title'
        type='text'
        placeholder='title'
        required={true}
      />
      {titleErrorComponents}
      {currentPostType === 'link' && (
        <>
          <FormField
            name='url'
            label='url'
            type='text'
            placeholder='url'
            required={true}
          />
          {urlErrorComponents}
        </>
      )}
      {currentPostType === 'text' && (
        <>
          <FormTextArea
            name='content'
            label='content'
            placeholder='content'
            required={true}
          />
          {contentErrorComponents}
        </>
      )}
      <FormButton label='create post' type='submit' />
    </Form>
  );
};

export default CreatePost;
