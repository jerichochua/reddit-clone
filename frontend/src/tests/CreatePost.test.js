import renderer from 'react-test-renderer';
import { AppProvider } from '../contexts/AppProvider';
import CreatePost from '../components/CreatePost/CreatePost';

it('renders correctly', () => {
  const tree = renderer.create(
    <AppProvider>
      <CreatePost />
    </AppProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
