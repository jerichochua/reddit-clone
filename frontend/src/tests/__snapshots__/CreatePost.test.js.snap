// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="form-container"
  style={
    Object {
      "maxWidth": "600px",
    }
  }
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-field"
    >
      <label>
        type
      </label>
      <select
        name="posttype"
        onChange={[Function]}
      >
        <option
          value="text"
        >
          text
        </option>
        <option
          value="link"
        >
          link
        </option>
      </select>
    </div>
    <div
      className="form-field"
    >
      <label>
        title
      </label>
      <input
        name="title"
        onChange={[Function]}
        placeholder="title"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-field"
    >
      <label>
        content
      </label>
      <textarea
        name="content"
        onChange={[Function]}
        placeholder="content"
        required={true}
        rows="6"
        value=""
      />
    </div>
    <button
      className="form-btn"
      type="submit"
    >
      create post
    </button>
  </form>
</div>
`;
